pkg/connection_pooler/connection_pooler.go

```diff
@@ -0,0 +1,938 @@
+package connection_pooler
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/r3labs/diff"
+	"github.com/sirupsen/logrus"
+	acidv1 "github.com/zalando/postgres-operator/pkg/apis/acid.zalan.do/v1"
+	"github.com/zalando/postgres-operator/pkg/cluster"
+	appsv1 "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/zalando/postgres-operator/pkg/resources"
+	"github.com/zalando/postgres-operator/pkg/util"
+	"github.com/zalando/postgres-operator/pkg/util/constants"
+	"github.com/zalando/postgres-operator/pkg/util/k8sutil"
+)
+
+const (
+	connectionPoolerContainer = "connection-pooler"
+	pgPort                    = 5432
+)
+
+// K8S objects that are belongs to a connection pooler
+type ConnectionPoolerObjects struct {
+	Deployment  *appsv1.Deployment
+	Service     *v1.Service
+	Name        string
+	ClusterName string
+	Namespace   string
+	logger      *logrus.Entry
+	// It could happen that a connection pooler was enabled, but the operator
+	// was not able to properly process a corresponding event or was restarted.
+	// In this case we will miss missing/require situation and a lookup function
+	// will not be installed. To avoid synchronizing it all the time to prevent
+	// this, we can remember the result in memory at least until the next
+	// restart.
+	LookupFunction bool
+}
+
+type SyncReason []string
+
+// no sync happened, empty value
+var NoSync SyncReason = []string{}
+
+// PostgresRole describes role of the node
+type PostgresRole string
+
+const (
+	// Master role
+	Master PostgresRole = "master"
```

Comment:
Role constants are already defined in [pkg/cluster/types.go](https://github.com/zalando/postgres-operator/blob/6a689cdc1c15bef7498cd29c0ad3e79a5f840996/pkg/cluster/types.go#L16). Why are they re-defined here ?